base62.go

```diff
@@ -25,28 +25,54 @@ func base2base(in []byte, inBase int, outBase int) []byte {
 		}
 
 		// this is really a prepend or remainder to result
-		res = append([]byte{byte(remainder)}, res...)
+		dst = append(dst, byte(remainder))
 		bs = quotient
 	}
 
-	return res
+	reverse(dst)
+	return dst
 }
 
-func encodeBase62(in []byte) string {
-	encoded := base2base(in, 256, 62)
-	for i, b := range encoded {
-		encoded[i] = base62Characters[b]
+func base2base(src []byte, inBase int, outBase int) []byte {
+	return appendBase2Base(nil, src, inBase, outBase)
+}
+
+func appendEncodeBase62(dst []byte, src []byte) []byte {
```

Comment:
The `Append` method is a very frequent optimization found in the standard library, which is why I don't feel too bad about it, and the upsides are great when it comes to performance.

Examples:
=> https://golang.org/pkg/strconv/#AppendBool
=> https://golang.org/pkg/time/#Time.AppendFormat
=> https://golang.org/pkg/math/big/#Int.Append

Yes, the consumer of `Append` has to understand that it's dealing with UTF-8 encoded data within the byte slice, but this is intended to be used as an optimization when building large strings, most of the code out there will keep using `String` as a way to get a string representation of the KSUID.

About the internal changes, I agree, it relies on the assumption that the character set we deal with here is all ASCII, which I think is OK since KSUIDs are represented in base 62. If that was to change we would break all code depending on it out there, and if we wanted to introduce alternative representation that use characters out of the ASCII space it would be a good time to revisit this code (although the likeliness makes me feel like the optimization is still worth it).